test/dynamo/test_subclasses.py

```diff
@@ -1304,74 +1372,53 @@ def fn(w):
         self.assertEqual(ref, res)
 
     def test_tensor_subclass_ctx_guards(self):
-        from torch.utils._python_dispatch import return_and_correct_aliasing
-
-        class SubclassTensor(torch.Tensor):
-            @staticmethod
-            def __new__(cls, a, constant):
-                shape = a.shape
-                kwargs = {}
-                kwargs["strides"] = a.stride()
-                kwargs["storage_offset"] = a.storage_offset()
-                kwargs["device"] = a.device
-                kwargs["layout"] = a.layout
-                kwargs["requires_grad"] = a.requires_grad
-                kwargs["dtype"] = a.dtype
-                out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)
-                return out
-
-            def __init__(self, a, constant):
-                self.a = a
-                self.constant = constant
+        x = CtxSubclassTensor(torch.ones(2), 3)
+        x2 = CtxSubclassTensor(torch.ones(2), 3)
+        x3 = CtxSubclassTensor(torch.ones(2), 4)
+        _check_recompiles(self, lambda x: x * x, (x,), (x2,), False)
+        _check_recompiles(self, lambda x: x * x, (x,), (x3,), True)
 
-            def __repr__(self):
-                a_repr = repr(self.a)
-                return f"SubclassTensor({a_repr})"
+    def test_tensor_subclass_ctx_custom_guards_override(self):
+        class CtxSubclassTensorCustomGuardFn(CtxSubclassTensor):
+            @classmethod
+            def __metadata_guard__(cls, orig_data, other):
```

Comment:
Now that we're guarding properly, we probably also want to add this guard function to `DTensor` ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L208)) or else we might see DTensor get a little slower with compile when graphs are small (DTensor has a lot of metadata, but in practice I don't think it wants to guard on all of it).

@wanchaol is it fair to say we probably dont' want to guard on the entire `DTensor._spec` (which contains a bunch of size/stride metadata that is redundant)? I would guess that we want the guard to at the very least only assert on `_spec.device_mesh` and `_spec.placements`,

```